test/test_torch.py

```diff
@@ -7735,6 +7735,72 @@ def test_helper(dim1, dim2, memory_format):
         test_helper((3, 3), (3, 3, 3, 3), torch.channels_last)
         test_helper((3, 3, 3), (3, 3, 3, 3, 3), torch.channels_last_3d)
 
+    def _check_dim_order_against_prim_impl(self, tensor):
+        import torch._prims_common as prim_utils
+
+        a = tuple(
+            prim_utils.compute_elementwise_output_logical_to_physical_perm(tensor, True)
+        )
+        b = tensor.dim_order()
+        self.assertSequenceEqual(a, b, seq_type=tuple)
+
+    def test_dim_order(self):
+        shape = (2, 3, 5, 7)
+
+        t = torch.empty(shape)
+        self.assertSequenceEqual(t.dim_order(), (0, 1, 2, 3), seq_type=tuple)
+        # transpose doesn't really change the underlying physical memory
+        # so execpting dim_order change to reflect that (like strides)
+        self.assertSequenceEqual(t.transpose(0, 1).dim_order(), (1, 0, 2, 3))
+        self._check_dim_order_against_prim_impl(t)
+
+        t = torch.empty(shape, memory_format=torch.channels_last)
+        self.assertSequenceEqual(t.dim_order(), (0, 2, 3, 1))
+        self._check_dim_order_against_prim_impl(t)
+
+        t = torch.empty((2, 3, 5, 7, 8), memory_format=torch.channels_last_3d)
+        self.assertSequenceEqual(t.dim_order(), (0, 2, 3, 4, 1))
+        self._check_dim_order_against_prim_impl(t)
+
+        for dim_order in itertools.permutations(range(4)):
+            self.assertSequenceEqual(
+                dim_order, torch.empty_permuted(shape, dim_order).dim_order()
+            )
+
+        self.assertRaises(
+            RuntimeError, lambda: torch.empty(shape).to_sparse().dim_order()
+        )
+        self.assertRaises(
+            RuntimeError, lambda: torch.empty(shape).to_mkldnn().dim_order()
+        )
+
+    def test_dim_order_ambiguous_cases(self):
+        K = 6
+        for ndim in range(K):
+            # generate a ndim array with ~same number of 1s and 2s e.g. [1, 1, 2, 2]
+            random_shape = random.sample([1, 2] * ndim, k=ndim)
+
+            for shape in set(itertools.permutations(random_shape)):
+
+                # cases with same strides, and size == 1
+                t = torch.empty(shape)
+                self._check_dim_order_against_prim_impl(torch.empty(shape))
```

Comment:
I would prefer if we had this test self contained, although assuming prim_impl is well tested this also seems reasonable. However, if we just direclty use prim_impl then these tests becomes moot or just test prim_impl.

Reply:
Since now we have two impls, the goal of this test is to make sure they align.